plugins/out_cloudwatch_logs/cloudwatch_api.c

```diff
@@ -0,0 +1,1072 @@
+
+        /* Get the record/map */
+        map = root.via.array.ptr[1];
+        map_size = map.via.map.size;
+
+        if (ctx->log_key) {
+            key_str = NULL;
+            key_str_size = 0;
+            check = FLB_FALSE;
+            found = FLB_FALSE;
+
+            kv = map.via.map.ptr;
+
+            for(j=0; j < map_size; j++) {
+                key = (kv+j)->key;
+                if (key.type == MSGPACK_OBJECT_BIN) {
+                    key_str  = (char *) key.via.bin.ptr;
+                    key_str_size = key.via.bin.size;
+                    check = FLB_TRUE;
+                }
+                if (key.type == MSGPACK_OBJECT_STR) {
+                    key_str  = (char *) key.via.str.ptr;
+                    key_str_size = key.via.str.size;
+                    check = FLB_TRUE;
+                }
+
+                if (check == FLB_TRUE) {
+                    if (strncmp(ctx->log_key, key_str, key_str_size) == 0) {
+                        found = FLB_TRUE;
+                        val = (kv+j)->val;
+                        ret = add_event(ctx, buf, stream, &val, &tms);
+                        if (ret < 0 ) {
+                            goto error;
+                        }
+                    }
+                }
+
+            }
+            if (found == FLB_FALSE) {
+                flb_plg_error(ctx->ins, "Could not find log_key '%s' in record",
+                              ctx->log_key);
+            }
+            else {
+                i++;
+            }
+            continue;
+        }
+
+        ret = add_event(ctx, buf, stream, &map, &tms);
+        if (ret < 0 ) {
+            goto error;
+        }
+        i++;
+    }
+    msgpack_unpacked_destroy(&result);
+
+    /* send any remaining events */
+    ret = send_log_events(ctx, buf, stream);
+    reset_flush_buf(ctx, buf, stream);
+    if (ret < 0) {
+        return -1;
+    }
+
+    /* return number of events */
+    return i;
+
+error:
+    msgpack_unpacked_destroy(&result);
+    return -1;
+}
+
+
+struct log_stream *get_dynamic_log_stream(struct flb_cloudwatch *ctx,
+                                          const char *tag, int tag_len)
+{
+    struct log_stream *new_stream;
+    struct log_stream *stream;
+    struct mk_list *tmp;
+    struct mk_list *head;
+    flb_sds_t name = NULL;
+    flb_sds_t tmp_s = NULL;
+    int ret;
+
+    name = flb_sds_create(ctx->log_stream_prefix);
+    if (!name) {
+        flb_errno();
+        return NULL;
+    }
+
+    tmp_s = flb_sds_cat(name, tag, tag_len);
+    if (!tmp_s) {
+        flb_errno();
+        flb_sds_destroy(name);
+        return NULL;
+    }
+    name = tmp_s;
+
+    /* check if the stream already exists */
+    mk_list_foreach_safe(head, tmp, &ctx->streams) {
+        stream = mk_list_entry(head, struct log_stream, _head);
+        if (strcmp(name, stream->name) == 0) {
+            flb_sds_destroy(name);
+            return stream;
+        } else {
```

Comment:
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#braces-usage-on-conditionals-loops-and-functions